crates/sui-core/src/consensus_adapter.rs

```diff
@@ -633,17 +656,22 @@ impl ConsensusAdapter {
         // this means we might be sending transactions from previous epochs to narwhal of
         // new epoch if we have not had this barrier.
         epoch_store
-            .within_alive_epoch(self.submit_and_wait_inner(transaction, &epoch_store))
+            .within_alive_epoch(self.submit_and_wait_inner(transactions, &epoch_store))
             .await
             .ok(); // result here indicates if epoch ended earlier, we don't care about it
     }
 
     #[allow(clippy::option_map_unit_fn)]
     async fn submit_and_wait_inner(
         self: Arc<Self>,
-        transaction: ConsensusTransaction,
+        transactions: Vec<ConsensusTransaction>,
         epoch_store: &Arc<AuthorityPerEpochStore>,
     ) {
+        assert!(transactions.len() > 0);
+        // TODO: to support soft bundle, we submit all transactions along with the first one.
+        // this is a dirty hack now as we did not check whether other transactions have already been submitted.
+        let all_transactions = &transactions;
+        let transaction = &transactions[0];
```

Comment:
Hmm, this seems like a potential issue. If I'm understanding this correctly, if the first transaction is submitted from elsewhere, then consensus adapter can decide that the entire batch is submitted even though only one transaction is been submitted.

This should be easily fixable by have `consensus_message_processed_notify` accept a slice of keys to wait for.